README.md
built_path/README.md

built_path/CHANGELOG.md
# CHANGELOG

## 0.1.0

- First release


built_path/test/src/simple_example.dart
// stub this out instead of trying to depend on Flutter for this repo
import 'package:built_path_annotations/built_path_annotations.dart';

part 'simple_example.svg_path.g.dart';

// stub, we're not referncing dart:ui here
enum PathFillType { nonZero, evenOdd }

// stub, we're not referencing dart:ui here
class Path {
  void lineTo(double x, double y) {}
  void moveTo(double x, double y) {}
  void cubicTo(
      double x, double y, double x2, double y2, double x3, double y3) {}
  void close() {}

  PathFillType fillType;
}

@SvgPath('M 10 10 L 20 20 z')
Path get path => _$path;

@SvgPath('M 250,75 L 323,301 131,161 369,161 177,301 z',
    fillRule: FillRule.evenOdd)
Path get star => _$star;

class Paths {
  @SvgPath('M 0 0 L 20 10 C 10 10 15 15 20 10 z')
  Path get anotherPath => _$Paths_anotherPath;
}


built_path/test/src/simple_example.svg_path.g.dart
// ignore_for_file: non_constant_identifier_names
// GENERATED CODE - DO NOT MODIFY BY HAND

part of 'simple_example.dart';

// **************************************************************************
// SvgPathGenerator
// **************************************************************************

Path __$path;
Path get _$path =>
    __$path ??
    (__$path = new Path()
      ..moveTo(10.0, 10.0)
      ..lineTo(20.0, 20.0)
      ..close());
Path __$star;
Path get _$star =>
    __$star ??
    (__$star = new Path()
      ..moveTo(250.0, 75.0)
      ..lineTo(323.0, 301.0)
      ..lineTo(131.0, 161.0)
      ..lineTo(369.0, 161.0)
      ..lineTo(177.0, 301.0)
      ..close()
      ..fillType = PathFillType.evenOdd);
Path __$Paths_anotherPath;
Path get _$Paths_anotherPath =>
    __$Paths_anotherPath ??
    (__$Paths_anotherPath = new Path()
      ..moveTo(0.0, 0.0)
      ..lineTo(20.0, 10.0)
      ..cubicTo(10.0, 10.0, 15.0, 15.0, 20.0, 10.0)
      ..close());


built_path/lib/src/path_generator.dart
import 'dart:async';

import 'package:analyzer/dart/constant/value.dart';
import 'package:analyzer/dart/element/element.dart';
import 'package:build/build.dart';
import 'package:built_path/src/flutter_path_proxy.dart';
import 'package:built_path_annotations/built_path_annotations.dart';
import 'package:path_parsing/path_parsing.dart';
import 'package:source_gen/source_gen.dart';

Builder svgPathSharedPartBuilder({String formatOutput(String code)}) {
  return new PartBuilder(
      <Generator>[new SvgPathGenerator()], '.svg_path.g.dart',
      formatOutput: formatOutput,
      header: '// ignore_for_file: non_constant_identifier_names\n'
          '// GENERATED CODE - DO NOT MODIFY BY HAND\n');
}

class SvgPathGenerator extends Generator {
  final TypeChecker _checker = const TypeChecker.fromRuntime(SvgPath);

  String _getFillRule(int index) {
    if (index == 0) {
      return '..fillType = PathFillType.nonZero';
    } else if (index == 1) {
      return '..fillType = PathFillType.evenOdd';
    }
    throw new StateError('Unhandled FillRule index $index');
  }

  void checkField(Element field, StringBuffer buffer, String friendlyName) {
    DartObject annotation = _checker.firstAnnotationOf(field);
    if (annotation == null && field is FieldElement) {
      annotation = _checker.firstAnnotationOf(field.getter);
    }
    if (annotation != null) {
      buffer.writeln('Path __\$$friendlyName;');
      buffer.writeln(
          'Path get _\$$friendlyName => __\$$friendlyName ?? (__\$$friendlyName =');
      final FlutterPathGenProxy proxy = new FlutterPathGenProxy();

      writeSvgPathDataToPath(
        annotation.getField('data').toStringValue(),
        proxy,
      );
      buffer.write(proxy);

      final int fillRuleIndex =
          annotation.getField('fillRule')?.getField('index')?.toIntValue();
      if (fillRuleIndex != null) {
        buffer.write(_getFillRule(fillRuleIndex));
      }
      buffer.writeln(');');
    }
  }

  @override
  FutureOr<String> generate(LibraryReader library, BuildStep buildStep) async {
    final StringBuffer buffer = new StringBuffer();
    for (Element el in library.allElements) {
      if (el is ClassElement) {
        for (FieldElement field in el.fields) {
          checkField(field, buffer, '${el.name}_${field.name}');
        }
      } else {
        checkField(el, buffer, el.name);
      }
    }
    return buffer.toString();
  }
}


built_path/example/lib/main.svg_path.g.dart
// ignore_for_file: non_constant_identifier_names
// GENERATED CODE - DO NOT MODIFY BY HAND

part of 'main.dart';

// **************************************************************************
// SvgPathGenerator
// **************************************************************************

Path __$MyHomePage_path;
Path get _$MyHomePage_path =>
    __$MyHomePage_path ??
    (__$MyHomePage_path = new Path()
      ..moveTo(20.0, 30.0)
      ..cubicTo(33.33333333333333, 13.333333333333332, 46.666666666666664,
          13.333333333333332, 60.0, 30.0)
      ..cubicTo(73.33333333333333, 46.666666666666664, 86.66666666666666,
          46.666666666666664, 100.0, 30.0));
Path __$MyHomePage_star;
Path get _$MyHomePage_star =>
    __$MyHomePage_star ??
    (__$MyHomePage_star = new Path()
      ..moveTo(250.0, 75.0)
      ..lineTo(323.0, 301.0)
      ..lineTo(131.0, 161.0)
      ..lineTo(369.0, 161.0)
      ..lineTo(177.0, 301.0)
      ..close()
      ..fillType = PathFillType.evenOdd);


built_path/example/lib/main.dart
import 'dart:ui' show Path;

import 'package:built_path_annotations/built_path_annotations.dart';
import 'package:flutter/material.dart';

part 'main.svg_path.g.dart';

void main() => runApp(new MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'Flutter Demo',
      theme: new ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: new MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({Key key, this.title}) : super(key: key);
  final String title;

  @SvgPath('M20,30 Q40,5 60,30 T100,30')
  // @SvgPath('M10 20 C 10 10 20 10 25 20 S 40 30 40 20')
  Path get path => _$MyHomePage_path;

  @SvgPath('M 250,75 L 323,301 131,161 369,161 177,301 z',
      fillRule: FillRule.evenOdd)
  Path get star => _$MyHomePage_star;

  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      appBar: new AppBar(
        title: new Text(title),
      ),
      body: new Column(
        children: <Widget>[
          new CustomPaint(painter: new PathPainter(path)),
          new CustomPaint(painter: new PathPainter(star, false)),
        ],
      ),
    );
  }
}

class PathPainter extends CustomPainter {
  PathPainter(this.path, [this.isStroke = true]);

  final Path path;
  final bool isStroke;

  @override
  void paint(Canvas canvas, Size size) {
    canvas.drawPath(
        path,
        new Paint()
          ..style = isStroke ? PaintingStyle.stroke : PaintingStyle.fill
          ..strokeWidth = 4.0);
  }

  @override
  bool shouldRepaint(PathPainter oldDelegate) => true;

  @override
  bool shouldRebuildSemantics(PathPainter oldDelegate) => false;
}
